### ISSUE TITLE
Type for validated DateTime

### ISSUE THREAD

**Issue by Razi91** [moment/luxon#1419]
**Is your feature request related to a problem? Please describe.**
With `@types/luxon` in version `3.3.0` it checks if the DateTime may be invalid, forcing to check if the result isn't null
```ts
const dateInJson = DateTime
    .local() // DateTime
    .startOf('day') // DateTime
    .plus({months: 3}) // DateTime
    .toJSON() // string | null
```

**Describe the solution you'd like**
A method like `.validated()` that either throw exception or return `Validated<DateTime>` skipping the possibility of `null` at `.toJSON()`

```ts
const dateInJson = DateTime
    .local() // DateTime
    .startOf('day') // DateTime
    .plus({months: 3}) // DateTime
    .validated() // Validated<DateTime> or throws exception
    .toJSON() // string
```
I'm not sure if `.local` shouldn't be already valid, `startOf` and `plus` shouldn't change it, therefore `.validated` shouldn't be even required here.

**Comment by zackdotcomputer** [comment]
+1 to this - the compiler should be informed that `DateTime.now().toISO()` will always be string, because `.now()` should always be valid.

**Comment by RuggeroCino** [comment]
Looks like a `throwOnInvalid` type settings has been introduced in version `3.3.0` (see [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64995)). As documented on the PR, the type issue can be fixed by overriding the `throwOnInvalid` type setting like:
```
// luxon.d.ts
import 'luxon';

declare module 'luxon' {
    interface TSSettings {
        throwOnInvalid: true;
    }
}
```
And setup Luxon to throw error on invalid dates as documented [here](https://github.com/moment/luxon/blob/master/docs/validity.md#throwoninvalid).

**Comment by zackdotcomputer** [comment]
I think the behavior @Razi91 has suggested would be an extension of `throwOnInvalid` - a way to ensure validity selectively rather than across the entire library. For example, there are some situations where you want to ensure validity absolutely (e.g. if loading a serialized date from a database) and some where you can tolerate failure (e.g. when accepting user input). As a workaround for now, I'll enable the `throwOnInvalid` behavior and use `try/catch` but being able to do this imperatively would feel more natural to me in TS.

As a side note, it would be great if the required snippets for enabling `throwOnInvalid` in both the code and typing were included in the documentation.

**Comment by shishkin** [comment]
I also stumbled upon this issue after [recent updates to typings](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64995). However I would like to suggest a different solution to the luxon library itself.

Instead of a global setting that one must not forget to set in every importing module luxon could export a "strict" version of API that will have `throwOnInvalid = true`. That way developer can chose which version to import and it will always be correctly typed.

I think this is better since it doesn't pollute API with extra `.validated()` method that I would need to call on every construction of luxon dates. It also allows for finer choice of where to use strict mode and where not, unlike a global nature of static settings (can lead to subtle bugs when set from different places) and global type override.

**Comment by zackdotcomputer** [comment]
I like the idea @shishkin, since yes there is risk that the Settings value gets set or unset somewhere (potentially even within a dependency!) and introduces bugs. But I would suggest you move it to its own issue so it doesn't get missed by the maintaining team. I'd say it's solving a distinct enough problem from the one we're looking at that both should stand in their own right and the library would be better if the team was able to incorporate both, not just one or the other.

**Comment by shishkin** [comment]
@zackdotcomputer you're right. Just created #1421.

**Comment by ctsstc** [comment]
Here for the following new type issue:

```typescript
const test: string = DateTime.now().toISODate()
// Type 'string | null' is not assignable to type 'string'.⏎ Type 'null' is not assignable to type 'string'.
```

**Comment by CarsonF** [comment]
I can't make any run time changes, but I've improved the types: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65078

```ts
DateTime.now().toISO(); // string
```
```ts
const input = DateTime.fromISO('...');

input.toISO(); // string | null

if (input.isValid) {
  input.toISO(); // string
  input.invalidReason; // null
} else {
  input.toISO(); // null
  input.invalidReason; // string
}
```
This is powered by a validity boolean generic. The default value is either valid or invalid (unless `throwOnInvalid` is enabled). To communicate a date was previously validated you can provide this value explicitly (`true` for valid, and `false` for invalid).
```ts
const now = DateTime.now();

function printDate(date: DateTime<true>): string {
  return date.toISO();
}
printDate(now);
```

**Comment by Andrew-Marks-Trisept** [comment]
> 
> After playing with this, I guess I need to handle nullability by adding a check and handling it, or defaulting it to some value.
> 
> ```ts
> const testOrNull: string | null = DateTime.now().toISODate()
> 
> if (testOrNull === null) {
>   // do something; return or throw
> }
> 
> const test: string = testOrNull // this will be valid now that we checked for the null scenario
> ```
> 
> 

I found a much simpler way to deal with this breaking change for my cases was to use the nullish coalescing operator, like this:

`const myString: string = myDate.toISODate() ?? '';`

**Comment by ctsstc** [comment]
@Andrew-Marks-Trisept that handles the typing, but likely you will not want to surface an empty string, most likely you'd want to handle it as an error or something rather than defaulting, in which case you'll end up with a typeguard/type narrowing instead.

**Comment by BrentSouza** [comment]
@CarsonF I think some of the type signatures are wrong in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65078

At least for [interval](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65078/files#diff-0e8e2c47dc4c85e1be2e301151ad8cb7b73533189f48903cddc50b5c113d38e7L96-R106)

```diff
     /**
     * Returns an error code if this Interval is invalid, or null if the Interval is valid
     */
-    get invalidReason(): string | null;
+    get invalidReason(): IfValid<string, null, IsValid>;

     /**
     * Returns an explanation of why this Interval became invalid, or null if the Interval is valid
     */
-    get invalidExplanation(): string | null;
+    get invalidExplanation(): IfValid<string | null, null, IsValid>;
```

I believe these should be: `IfValid<null, string, IsValid>`. Do you agree?

I haven't looked further beyond that, but just noticed this error pop up in my code when trying to narrow types:
```ts
if (interval.isValid) {
  // do stuff
} else {
  const error: string = interval.invalidExplanation; //  Type 'null' is not assignable to type 'string'.
}
```

**Comment by CarsonF** [comment]
@BrentSouza Yes, you're right. My bad. Looks like `DateTime` was fixed in #67756. Do you want to PR a fix?

**Comment by BrentSouza** [comment]
Sure. I'll open a PR shortly

**Comment by ptandler** [comment]
I also just stumbled across this. I thought I could use a type guard to convert the type of a given DateTime from `DateTime<boolean>` to `DateTime<true>` like this:

```ts
export function isDateTimeValid(
  dateTime: DateTime<boolean> | undefined | null,
): dateTime is DateTime<true> {
  return dateTime?.isValid ?? false;
}
```

Which seems to work fine **if dateTime is of type `DateTime<true> | DateTime<false>`**:

```ts
    const dateTime = DateTime.fromISO('2222-12-31T23:59:59.999Z');

    if (isDateTimeValid(dateTime)) {
      const iso = dateTime.toISO();
      console.log(iso); // typeof iso is always string
    }

```

But it does not work if dateTime is of type `DateTime<boolean>`: 

```ts
    const dateTime1: DateTime<boolean> = DateTime.fromISO('2222-12-31T23:59:59.999Z');

    if (isDateTimeValid(dateTime1)) {
      const iso = dateTime1.toISO();
      console.log(iso); // typeof iso is string | null
    }

```

This seems to be a limitation of current TypeScript. But, if I ensure that I use union types in all of my methods that return DateTime, this is a step forward :slightly_smiling_face:

TypeScript does not support turning the `DateTime.isValid` directly into a type guard like this, right?

**Comment by CarsonF** [comment]
@ptandler No TS cannot use a getter as a type guard. Since the JS src is a getter, that is a constraint we cannot control.

TS can narrow unions this way though. Which is how I've tried to type it.
`fromISO` (like many) returns a union.
```ts
fromISO(iso: string): DateTimeMaybeValid
// aka
fromISO(iso: string): DateTime<true> | DateTime<false>
```

So letting the type be inferred can narrow it.
```ts
 const dateTime1 = DateTime.fromISO('2222-12-31T23:59:59.999Z');
if (dateTime1.isValid) {
  const iso: string = dateTime1.toISO();
}
```

This doesn't work when declaring a parameter though.
```ts
foo(dt: DateTime)
foo(dt: DateTime<boolean>)
```
Since `boolean` satisfies both `true` and `false` checks it is impossible to narrow the method return types.

I had considered trying to export `DateTime` as this type union. But I gave up at the time because there's also static methods to consider.

Something like
```ts
export type DateTime = DateTimeValid | DateTimeInvalid;

class DateTimeValid { ... };
class DateTimeInvalid { ... };
```
I think this might break `instanceof` checks though. And overall strays very far from the JS src.

So again, this is a really hard problem for TS to express safely.

I'll admit that I don't spend too much time with this in my own usage, because I always enable throwing on errors instead of carrying the errors forward as values.
```ts
import { Settings } from 'luxon';

Settings.throwOnInvalid = true;
declare module 'luxon' {
  interface TSSettings {
    throwOnInvalid: true;
  }
}
```

Assuming all values are valid is much easier, and the few places where there needs to be input checks those can be wrapped in try/catch.
